lib/grandpa/grandpa.go

```diff
@@ -449,188 +424,300 @@ func (s *Service) primaryBroadcastCommitMessage() {
 // playGrandpaRound executes a round of GRANDPA
 // at the end of this round, a block will be finalised.
 func (s *Service) playGrandpaRound() error {
-	logger.Debugf("starting round %d with set id %d",
-		s.state.round, s.state.setID)
+	currentRound := s.state.round
+	currentSetID := s.state.setID
+
 	start := time.Now()
 
-	ctx, cancel := context.WithCancel(s.ctx)
-	defer cancel()
+	logger.Debugf("starting round %d with set id %d",
+		currentRound, currentSetID)
 
-	isPrimary, err := s.handleIsPrimary()
-	if err != nil {
-		return err
-	}
+	cancel := make(chan struct{})
+	defer close(cancel)
 
-	logger.Debug("receiving pre-vote messages...")
-	go s.receiveVoteMessages(ctx)
-	time.Sleep(s.interval)
+	roundActions := s.finalizationEngine(cancel)
```

Comment:
I think we need to know that all `finalizationEngine` goroutines have actually exited if we actually close `cancel`.  It may make sense to create a `finalizationEngine` type that implements the service interface that I suggested:
```
// Service is a generalized interface for a long running asynchronous service
type Service interface {
	// Start returns an error channel that is only written to if err == nil.  It is implied that
	// the service is ready if the returned err == nil.
	Start() (errChan chan error, err error)
	// Stop will shutdown the service.  The errChan returned in Start() is expected to be closed.
	// Packages that implement the Service interface should implement a timeout to ensure Stop()
	// does not block indefinitely.
	Stop() (err error)
}```

That way we can ensure that no goroutine is left dangling before proceeding to the next grandpa round.